### ISSUE TITLE
Persistent notifications after reboot

### ISSUE THREAD

**Issue by mymike00** [ubports/ubuntu-touch#354]
- Device: krillin (bq E4.5)
- Channel: rc
- Build: r10

## Steps to reproduce

1. wait for some notifications or create them with something like NPost
2. reboot the phone

## Expected behavoir

I expect to see the notifications I had before rebooting  in the indicator

## Actual behavior

Notifiactions I got before are no more visible in the indicator

**Comment by dpniel** [comment]
There is also the possibilty of the push client having already delivered the notifications to the receiving app and flushed it's buffer but the notifications still remain in the indicator because the user/app is yet to clear them.

**Comment by dpniel** [comment]
@mardy that would mean the push-client would also need to restore them as well, otherwise you have lost the `message` portion of the push message. The app receives a different payload to what the user sees in the indicator. So which would be the source of truth? What happens if the push-client has already delivered the message payload to the app but the notification is still in the indicator? Should it still be restored even though the app has already received it. 

I think NPost is a special case here as it will immediately receive the message payload on posting the notification to the postal service, and uses the ability to not clear the notification as a way to make it persist in the tray. But in the eyes of the push-client it has been "delivered"

**Comment by mardy** [comment]
Oh, I see, this should be handled by the push client, but getting it right might not be as simple as I thought.

**Comment by dpniel** [comment]
Yeah it's a tricky one :-) That's why I lean towards the current approach being correct in how it handles it. It would be nice to have though...

Ideally i think on reboot an app should do the checks or poke a server to see if it has any unhandled messages and repost notifications if appropriate. But the lifecycle prevents that.

**Comment by dpniel** [comment]
I think a polld plugin would provide quite a nice solution for NPost. As it could replay it's stored notifications by comparing message tags from the postal service with it's internal list, and remove from the internal list as they get activated by the message indicator.

**Comment by mymike00** [comment]
mmh, I'm not understanding all you're speaking, but I would like persistent notifications as my touchscreen usually gets mad and it start tapping randomly on the screen and other times the touchscreen doesn't work at all because of the screen broken and I have to restart the phone... this could happens lots of times a day and when it happens I lost all the notifications.
If it could be easier, notifications restored at startup could not have all the information (like the `message` @dpniel were talking about, and I don't know what is it tbh...): just the app logo and the title? or something like this, just to remember I had a notification from that app and what is about... It would nicer than now...

**Comment by dpniel** [comment]
@mymike00 i think restoring only part of the notification is the wrong way to go. What if the receiving app depends on the data in the `message` part you say should get dropped as you only care about the visual part? What if it's a non visual notification? There is alot of edge cases to think about and the desired result could be different for each app. The app or service is the only one who knows the actual data that should be re-displayed or re-sent.

The way i think it should work is:

1. Phone boots
2. App helper runs (possibly as a polld plugin)
3. App checks either an internal data store and reposts any notifications if it's a local poll service, or contacts a server to resend notifications for any unhandled data.

The push-client and messaging indicator should be a passive service that just displays data and passes the data on to the receiving app when it can.

**Comment by dobey** [comment]
The push client has nothing to do with the UI. We shouldn't run all the helpers on phone boot to check for notifications either.

Really what we need here is an entire rearchitecting of notifications. We had some plans to do this before Canonical dropped Unity, but never got the time to do it. Ideally this would include getting rid of indicator-messages as well.
